### PR TITLE
feat: support vllm in controller

### PR DESCRIPTION
**Pull request by zhuangqh**
**Reason for Change**:

- support vllm runtime deployments
- add a feature gate to enable vllm runtime
- able to select runtime by annotation tag

**Requirements**

- [x] added unit tests and e2e tests (if applicable).

### CODE REVIEW

File: docker/presets/models/tfs/Dockerfile [hunk 1]
```diff
@@ -5,31 +5,39 @@ ARG MODEL_TYPE
 ARG VERSION
```
Comment (zhuangqh) [inline docker/presets/models/tfs/Dockerfile:5]:
```
.
├── chat_templates
│   ├── alpaca.jinja
│   ├── amberchat.jinja
│   ├── chatml.jinja
│   ├── chatqa.jinja
│   ├── falcon-instruct.jinja
│   ├── gemma-it.jinja
│   ├── llama-2-chat.jinja
│   ├── llama-3-instruct.jinja
│   ├── mistral-instruct.jinja
│   ├── openchat-3.5.jinja
│   ├── phi-3-small.jinja
│   ├── phi-3.jinja
│   ├── saiga.jinja
│   ├── solar-instruct.jinja
│   ├── vicuna.jinja
│   └── zephyr.jinja
├── tfs
│   ├── cli.py
│   ├── dataset.py
│   ├── fine_tuning.py
│   ├── inference-requirements.txt
│   ├── inference_api.py
│   ├── metrics_server.py
│   ├── parser.py
│   ├── tuning-requirements.txt
│   └── weights -> /workspace/weights
├── version.txt
├── vllm
│   ├── inference-requirements.txt
│   ├── inference_api.py
│   └── weights -> /workspace/weights
└── weights
```

File: pkg/model/interface.go [hunk 2]
```diff
@@ -13,23 +15,131 @@ type Model interface {
...
+	ModelRunParams     map[string]string // Parameters for running the model training/inference.
+}
+
+type VLLMParam struct {
+	BaseCommand        string            // The initial command (e.g., 'torchrun', 'accelerate launch') used in the command line.
+	DistributionParams map[string]string // Parameters for distributed inference.
+	ModelRunParams     map[string]string // Parameters for running the model training/inference.
+}
+
+func (p *PresetParam) DeepCopy() *PresetParam {
```
Comment (zhuangqh) [inline pkg/model/interface.go:69]: we may update some params according to the node counts. Thus, we must deepcopy it at first.

File: pkg/model/interface.go [hunk 3]
```diff
@@ -13,23 +15,141 @@ type Model interface {
...
-	DiskStorageRequirement        string            // Disk storage requirements for the model.
-	GPUCountRequirement           string            // Number of GPUs required for the Preset. Used for inference.
-	TotalGPUMemoryRequirement     string            // Total GPU memory required for the Preset. Used for inference.
-	PerGPUMemoryRequirement       string            // GPU memory required per GPU. Used for inference.
-	TuningPerGPUMemoryRequirement map[string]int    // Min GPU memory per tuning method (batch size 1). Used for tuning.
-	TorchRunParams                map[string]string // Parameters for configuring the torchrun command.
-	TorchRunRdzvParams            map[string]string // Optional rendezvous parameters for distributed training/inference using torchrun (elastic).
-	BaseCommand                   string            // The initial command (e.g., 'torchrun', 'accelerate launch') used in the command line.
-	ModelRunParams                map[string]string // Parameters for running the model training/inference.
+	Tag             string // The model image tag
```
Comment (ishaansehgal99) [inline pkg/model/interface.go:31]: does this tag field get used, why include

Comment (zhuangqh) [reply pkg/model/interface.go:31]: nope. just alter this line of code from L34

**Comment by zhuangqh** [comment]
ready @Fei-Guo

**Review by Fei-Guo — APPROVED** [review]

**Pull request merged**